adr/0016-hooks.md

```diff
@@ -0,0 +1,113 @@
+	Lifecycle    HookLifecycle          `json:"lifecycle" jsonschema:"description=Lifecycle of the hook"`
+	HookData     map[string]interface{} `json:"hookData" jsonschema:"description=Generic data map used for the hook. The data is obtained from a secret in the Zarf namespace"`
+	OCIReference string                 `json:"ociReference" jsonschema:"description=Optional OCI reference to the hook image to run"`
+}
+```
+
+Example Secret Data:
+```yaml
+hookName: ecr-repository
+internal: true
+lifecycle: before-component
+hookData:
+  registryURL: public.ecr.aws/abcdefg/zarf-ecr-registry
+  region: us-east-1
+  repositoryPrefix: ecr-zarf-registry
+```
+
+For this solution, hooks have to be 'installed' onto a cluster before they are used. When Zarf is deploying a package onto a cluster, it will look for any secrets with the `zarf-hook` label in the `zarf` namespace.  If hooks are found, Zarf will run any 'package' level hooks before deploying a component and run any 'component' level hook for each component that is getting deployed. The hook lifecycle options will be before/after a package deployment and before/after a component deployment.
+ - NOTE: The order of hook execution is nearly random. If there are multiple hooks for a lifecycle there is no guarantee that they will be executed in a certain order.
+ - NOTE: The `package` lifecycle might be changed to a `run-once` lifecycle. This would benefit packages that don't have kube context information when the deployment starts.
+
+Zarf hooks will have two forms of execution via `Internal` and `External` hooks.
+### Internal Hooks
+Internal hooks will be hooks that are built into the Zarf CLI and run internal code when executed. The logic for these hooks would be built into the Zarf CLI and would be updated with new releases of the CLI.
+
+### External Hooks
+There are a few approaches for external hooks.
+1. Have the hook metadata reference an OCI image that is downloaded and run.
+ - The hook metadata can reference the shasum of the image to ensure the image is not tampered with.
+ - We can pass metadata from the secret to the image.
+2. Have the hook metadata reference an image/endpoint that we call via a gRPC call.
+ - This would require a lot of consideration to security since we will be executing code from an external source.
+3. Have the hook metadata contain a script or list of shell commands that can get run.
+ - This would be the simplest solution but would require the most work from the hook creator. This also has the most potential security issues.
+
+
+
+**PROS**
+ - Implementing Hooks internally means we don't have to deal with any bootstrapping issues.
+ - Internally managed hooks can leverage Zarf internal code.
+
+**CONS**
+ - Since 'Internal' hooks are built into the CLI, the only way to get updates for the hook is to update the CLI.
+ - External hooks will have a few security concerns that we will have to work through.
+
+
+
+### Pepr
```

Comment:
We may want to just say "Webhooks" - this _could_ be Pepr or it could be some other random API.  It might also be worth showing what the secrets / lifecycle parts would look like for this too (i.e. instead of `OCIReference` it would show a `URL` and / or a `Selector` of some kind)